### ISSUE TITLE
error in parse MessageID

### ISSUE THREAD

**Issue by talenor** [jstedfast/MailKit#224]
Don't parse MessageId without '<', '>' on Fetch

Sometimes there are records of this type:
Message-ID: faff5b13bff19cf16f0fbf46edb721e3@zohomail.com

**Comment by jstedfast** [comment]
You realize that's invalid, right? The Message-Id value MUST MUST MUST be between <>'s

**Comment by talenor** [comment]
but in real life i receive mails Message-ID without <>

another sample with full headers

Received: from WIN-ST-MBX-04.store.local (172.16.20.90) by
 WIN-ST-MBX-03.store.local (172.16.20.131) with Microsoft SMTP Server (TLS) id
 15.0.1076.9 via Mailbox Transport; Thu, 6 Aug 2015 13:06:00 +0300
Received: from WIN-ST-EDGE-01.store.local (172.16.20.13) by
 WIN-ST-MBX-04.store.local (172.16.20.90) with Microsoft SMTP Server (TLS) id
 15.0.1076.9; Thu, 6 Aug 2015 13:06:00 +0300
Received: from p3nlsmtp15.shr.prod.phx3.secureserver.net (72.167.234.240) by
 mx1.mail.oversun.ru (188.127.232.5) with Microsoft SMTP Server id 14.3.210.2;
 Thu, 6 Aug 2015 13:06:00 +0300
Received: from p3nlwpweb137.shr.prod.phx3.secureserver.net ([50.62.208.129])
	by p3nlsmtp15.shr.prod.phx3.secureserver.net with 	id 1N5x1r01w2o2uXE01N5xDx;
 Thu, 06 Aug 2015 03:05:57 -0700
Received: from mx1.mail.oversun.ru (localhost [127.0.0.1])	by
 p3nlwpweb137.shr.prod.phx3.secureserver.net (8.14.4/8.12.11) with SMTP id
 t76A5uiv362830	for <info@oversun.ru>; Thu, 6 Aug 2015 03:05:56 -0700
Date: Thu, 6 Aug 2015 03:05:56 -0700
From:
	=?utf-8?B?0JHQuNC30L3QtdGBINGG0LXQvdGC0YAg0YDRj9C00L7QvCDRgSDQktCw0LzQuA==?=
	<mhygQvUq@aol.com>
To: <info@oversun.ru>
Subject: =?utf-8?B?0J/QviDQv9C+0LLQvtC00YMg0LDRgNC10L3QtNGLINGB0LrQu9Cw0LTQsCzQvtGE0LjRgdCwLCDQv9GA0L7QuNC30LLQvtC00YHRgtCy0LXQvdC90L7Qs9C+INC4INGC0L7RgNCz0L7QstC+0LPQviDQv9C+0LzQtdGJ0LXQvdC40LkuINCR0LXQtyDQv9C+0YHRgNC10LTQvdC40LrQvtCyLg==?=
Message-ID: d5a54866851d9c1e1e0ffd6618ff9fab@aol.com
MIME-Version: 1.0
Content-Type: text/html; charset="UTF-8"
Return-Path: mhygqvuq@aol.com
X-MS-Exchange-Organization-MessageDirectionality: Incoming
X-MS-Exchange-Organization-PRD: aol.com
Received-SPF: SoftFail (WIN-ST-EDGE-01.store.local: domain of transitioning
 mhygQvUq@aol.com discourages use of 72.167.234.240 as permitted sender)
X-MS-Exchange-Organization-Antispam-Report: v=2.1 cv=ELA/jTpC c=1 sm=1 tr=0
 a=4LbsiieQo/6nd9RHBuh6nQ==:117 a=3oc9M9_CAAAA:8 a=TZb1taSUAAAA:8
 a=SVDnmV8JAAAA:8 a=jPJDawAOAc8A:10 a=IkcTkHD0fZMA:10 a=uRRa74qj2VoA:10
 a=SDZOPyz1s2nSdqcScU4A:9 a=P_XQIPuGG9yWRcSb:21
 a=QEXdDO2ut3YA:10;OrigIP:72.167.234.240;SCL:-1
X-MS-Exchange-Organization-AVStamp-Mailbox: MSFTFF;1;0;0 0 0
X-MS-Exchange-Organization-SCL: -1
X-MS-Exchange-Organization-SenderIdResult: SOFTFAIL
X-MS-Exchange-Organization-Network-Message-Id: bc189c2b-f836-41c4-d1b8-08d29e46a148
X-MS-Exchange-Organization-AVStamp-Enterprise: 1.0
X-Auto-Response-Suppress: DR, OOF, AutoReply
X-MS-Exchange-Organization-AuthSource: WIN-ST-EDGE-01.store.local
X-MS-Exchange-Organization-AuthAs: Anonymous

**Comment by jstedfast** [comment]
Have you submitted a bug report against the mail software that created such messages?

**Comment by andylholloway** [comment]
I'm also experiencing a similar problem but with colons ':' in the Message-ID value e.g.
If we receive a mail with the following Message-ID: <comanyname-http://company.com>
and pass the message into MimeMessage.Load(), the MimeMessage.MessageId will be null.

Would you expect this result?

**Comment by jstedfast** [comment]
Yes, I would expect that.

```
   msg-id          =   [CFWS] "<" id-left "@" id-right ">" [CFWS]

   id-left         =   dot-atom-text / obs-id-left

   id-right        =   dot-atom-text / no-fold-literal / obs-id-right

   no-fold-literal =   "[" *dtext "]"

   atext           =   ALPHA / DIGIT /    ; Printable US-ASCII
                       "!" / "#" /        ;  characters not including
                       "$" / "%" /        ;  specials.  Used for atoms.
                       "&" / "'" /
                       "*" / "+" /
                       "-" / "/" /
                       "=" / "?" /
                       "^" / "_" /
                       "`" / "{" /
                       "|" / "}" /
                       "~"

   atom            =   [CFWS] 1*atext [CFWS]

   dot-atom-text   =   1*atext *("." 1*atext)

   dot-atom        =   [CFWS] dot-atom-text [CFWS]

   specials        =   "(" / ")" /        ; Special characters that do
                       "<" / ">" /        ;  not appear in atext
                       "[" / "]" /
                       ":" / ";" /
                       "@" / "\" /
                       "," / "." /
                       DQUOTE
```

Notice that :'s are illegal.

What software are you using that sends http urls in a Message-ID header? Was it written by people that had even the slightest clue what they were doing? Clearly the answer is no. They obviously slapped together something vaguely message-like without ever reading the specifications.

My suggestion to you is to do this:

```csharp
var msgid = message.Headers[HeaderId.MessageId].Value.Trim ();
```

I'm not going to add support for parsing the kind of garbage that you get in that header.

**Comment by andylholloway** [comment]
Thanks for the quick reply and the suggestion on how to retrieve the message id.
I'm not involved with the system or software used to generate the mail with the invalid message id. Our smtp server can receive mail from anywhere and it is not unusual to receive mail that has an invalid or empty Message-ID.

Within our system we do the following:

private static void EnsureMessageHasAnId(MimeMessage mail)
        {
            if (string.IsNullOrWhiteSpace(mail.MessageId))
            {
                mail.MessageId = MimeUtils.GenerateMessageId();
            }
        }

We are currently using MimeKit v1.0.10.0.

**Comment by jstedfast** [comment]
There have been a lot of MimeKit fixes since 1.0.10 :)
